client/src/boot/BootRoutes.js

```diff
@@ -140,8 +158,79 @@ class BootRoutes {
       });
     });
 
+    // Legacy support – make sure we don't overwrite the `onbeforeunload`
+    // handler in `LeftAndMain.EditForm.js` but run it after the new handler.
+    const currBeforeUnload = window.onbeforeunload;
+    window.onbeforeunload = () => {
+      if (this.shouldConfirmBeforeUnload()) {
+        return i18n._t('Admin.CONFIRMUNSAVEDSHORT', 'WARNING: Your changes have not been saved.');
+      }
+
+      if (typeof currBeforeUnload === 'function') {
+        return currBeforeUnload();
+      }
+
+      return null;
+    };
+
     pageRouter.start();
   }
+
+  /**
+   * Return `true` if there are forms with unsaved changes. Return `false`
+   * otherwise.
+   *
+   * @return {Boolean}
+   */
+  shouldConfirmBeforeUnload() {
+    const state = this.store.getState();
+    const unsaveds = state.unsavedForms || {};
+    let ret = false;
+
+    if (unsaveds && Object.keys(unsaveds).length > 0) {
+      ret = true;
+    }
+
+    return ret;
+  }
+
+  /**
+   * Return `true` if the route the user is navigating away from contains forms
+   * with unsaved changes. Return `false` otherwise.
+   *
+   * @return {Boolean}
+   */
+  shouldConfirmBeforeRoute() {
+    const state = this.store.getState();
+    const unsaveds = state.unsavedForms || {};
+    const locationBeforeTransitions = state.routing.locationBeforeTransitions;
+    const pathname = locationBeforeTransitions && locationBeforeTransitions.pathname;
+    let ret = false;
```

Comment:
@phalkunz for future reference, a meaningful variable name would be helpful, e.g. `shouldConfirm`